Eliminated Psalm baseline, migrated from `webmozart/assert` to `azjezz/psl`

Ocramius opened the pull request:
This change is quite massive, and rewrites most of the library so that it is more type-safe, especially when it comes to large JSON structure traversal.

`composer/composer:^1` manifest parsing is now gone (not supported anyway, since we depend on `composer-runtime-api:^2`).

We migrated from `webmozart/assert` to the incredibly more powerful `azjezz/psl` dependency, which gives us massive code reduction, and improved type guarantees.

Most tests were also simplified a lot, and some (outdated, due to older PHP versions targeted) are gone.

Test metadata has also been rewritten, and PHPUnit has been tuned to be much stricter.

`symfony/console:^7.2.1` has also been bumped, meanwhile.

Ocramius committed:
Eliminated Psalm baseline, migrated from `webmozart/assert` to `azjezz/psl`

This change is quite massive, and rewrites most of the library so that it is
more type-safe, especially when it comes to large JSON structure traversal.

`composer/composer:^1` manifest parsing is now gone (not supported anyway, since
we depend on `composer-runtime-api:^2`).

We migrated from `webmozart/assert` to the incredibly more powerful `azjezz/psl`
dependency, which gives us massive code reduction, and improved type guarantees.

Most tests were also simplified a lot, and some (outdated, due to older PHP
versions targeted) are gone.

Test metadata has also been rewritten, and PHPUnit has been tuned to be much
stricter.

`symfony/console:^7.2.1` has also been bumped, meanwhile.

Ocramius committed:
Raising PHPStan analysis level to `max`

This leads to some minor tradeoffs between `non-empty-string` vs `string`
declarations, and required us to remove an `@pure` declaration, but is
otherwise a good improvement overall.

Ocramius committed:
Moved complex JSON structure declarations to the internal `JsonLoader`

This type is relatively safe to keep inside this package (only), and
its only responsibility is really behaving as anti-corruption layer
with `composer.json`, `composer.lock` and `installed.json`, so
this is a good fit.

Ocramius committed:
Install `intl`, `bcmath` and `mbstring` dependencies in CI

These are minor dependency upgrades that are well worth having, for
the reduced code complexity in our domain.

Ocramius committed:
Corrected file reference in `PharTest`, which didn't run locally

Ocramius committed:
Run mutation tests only on the latest PHP version

These are slow and wasteful: not worth burning CPUs for no reason.

Ocramius committed:
Skip running PHPUnit with lowest dependencies on PHP 8.4

Many libraries, such as Symfony, don't understand semantic versioning
well, and tend to release software that is incompatible with the next
major version of a dependency (in this case `php`, which does not
follow SemVer).

This kind of botched architecture by Symfony leads to dependencies having
broken behavior in oldstable releases.

This is effectively a choice by the upstream maintainer (Symfony) to
leave this sort of problem to the consumer (in this case, us), who then
has to restrict ranges more aggressively.

In this case, the failures are due to transitive dependencies leading to
symfony, so introducing a `conflict` worsens the situation for our own
consumers/

Ocramius commented:
@maglnet since I'm touching a lot of the project, you might want to take a glance here, especially to see how I'm ripping out pieces to use `azjezz/psl` instead, which is a fabulous library :D

fredden commented on .github/workflows/phpunit.yml:
```diff
@@ -25,6 +25,9 @@ jobs:
                 operating-system:
                     - "ubuntu-latest"
                     - "windows-latest"
+                exclude:
+                  - dependencies: "lowest"
+                    php-version: "8.4"
```
Why are we excluding this combination from the tests?

Ocramius replied:
I added a dedicated commit about it, but basically, symfony still pretends to release software with deathstar-alike dependency ranges

fredden replied:
Thanks for this. I didn't look through the individual commits. Can we add this as a code comment so that it's visible for future contributors?

Ocramius replied:
I'm unsure about that: the decision may be made to re-add them later, for other reasons, so I don't think we should put a doorstopper here

fredden commented on .github/workflows/infection.yml:
```diff
@@ -17,8 +17,6 @@ jobs:
         dependencies:
           - "locked"
         php-version:
-          - "8.2"
-          - "8.3"
```
Why are we removing these PHP versions from the matrix? They are still listed in `composer.json` as being supported.

Ocramius replied:
Very wasteful to run mutation tests for each version, while we only care about the software being well tested (for software reliability, normal tests suffice)

fredden replied:
Thanks. This makes sense. Sorry for the noise.

fredden commented on src/ComposerRequireChecker/Cli/CheckCommand.php:
```diff
@@ -225,18 +220,15 @@ static function (string $unknownSymbol) use ($guesser): array {
-     */
+    /** @throws InvalidJson */
     private function getCheckOptions(InputInterface $input): Options
     {
-        $configFileArgument = $input->getOption('config-file');
-        assert(is_string($configFileArgument) || $configFileArgument === null);
+        $configFileArgument = Type\union(Type\null(), Type\non_empty_string())
+            ->coerce($input->getOption('config-file'));
 
```
This looks like it will now potentially throw `Type\Exception\CoercionException` for an empty string.

Ocramius replied:
Yep: an  empty file is certainly not valid

fredden commented on src/ComposerRequireChecker/Cli/CheckCommand.php:
```diff
@@ -225,18 +220,15 @@ static function (string $unknownSymbol) use ($guesser): array {
-            if ($fileName === false) {
+            $fileName = Filesystem\canonicalize($configFileArgument);
+            if ($fileName === null) {
                 throw new InvalidArgumentException(
```
Does this need to be listed in the `@throws` docblock?

Ocramius replied:
Possibly: I added `@throws` in an age when the IDE was really pedantic about it.

I don't believe them to be as important anymore, but if you believe so, we can tweak SA tools to check these

fredden replied:
I think we should either use the tag or not. I'm happy with the tags going away / being removed.

Ocramius replied:
I'd say that they are relevant for things where we explicitly also test for them

fredden commented on test/ComposerRequireCheckerTest/Cli/OptionsTest.php:
```diff
@@ -63,9 +62,22 @@ public function testOptionsFileRepresentsDefaults(): void
         $fileContent = file_get_contents(__DIR__ . '/../../../data/config.dist.json');
         $this->assertNotFalse($fileContent);
 
-        $optionsFromFile = new Options(
-            json_decode($fileContent, true),
-        );
+        $optionsFromFile = new Options([
+            'symbol-whitelist'    => [],
+            'php-core-extensions' => [
+                'Core',
+                'date',
+                'json',
+                'hash',
+                'pcre',
+                'Phar',
+                'Reflection',
+                'SPL',
+                'random',
+                'standard',
+            ],
+            'scan-files'          => [],
+        ]);
 
         $this->assertEquals($options, $optionsFromFile);
     }
```
`$fileContent` is no longer used in this test. We are now comparing `new Options()` with `new Options( $some_fixed_content )`. The file isn't used at all here any longer. Perhaps this test should be renamed or rewritten?

Ocramius replied:
Yep, renaming is a good idea

Ocramius replied:
8852124bd26ad5d74fd33eb3a99a1596e1a08cb8

fredden commented on test/ComposerRequireCheckerTest/DefinedSymbolsLocator/LocateDefinedSymbolsFromComposerRuntimeApiTest.php:
```diff
@@ -22,19 +23,23 @@ protected function setUp(): void
+        self::assertEmpty(($this->locator)(Json\typed(
+            $composerJson,
+            JsonLoader::composerDataType(),
+        )));
     }
 
     /** @dataProvider provideComposerJsonWithSuitableComposerRuntimeApi */
+    #[DataProvider('provideComposerJsonWithSuitableComposerRuntimeApi')]
```
Do we need both of these comments? Can we remove the docblock tag in favour of the attribute?

Ocramius replied:
I kept the old comments due to `psalm/plugin-phpunit` relying on them, but I'm working ti fix the upstream package

fredden commented on test/ComposerRequireCheckerTest/FileLocator/LocateAllFilesByExtensionTest.php:
```diff
@@ -96,6 +97,7 @@ public function testLocateWithDirectoryBlackList(): void
      *
      * @dataProvider provideBlacklists
      */
+    #[DataProvider('provideBlacklists')]
```
Do we need both of these comments? Can we remove the docblock tag in favour of the attribute?

Ocramius replied:
https://github.com/maglnet/ComposerRequireChecker/pull/564#discussion_r1946739239

fredden commented on test/ComposerRequireCheckerTest/FileLocator/LocateComposerPackageSourceFilesTest.php:
```diff
@@ -148,6 +152,7 @@ public function testFromPsr4WithNestedDirectoryAlternativeDirectorySeparator():
      *
      * @dataProvider provideExcludePattern
      */
+    #[DataProvider('provideExcludePattern')]
```
Do we need both of these comments? Can we remove the docblock tag in favour of the attribute?

Ocramius replied:
https://github.com/maglnet/ComposerRequireChecker/pull/564#discussion_r1946739239

Ocramius committed:
Applied CS fixes due to sorted imports after a rebase

Ocramius committed:
Adjusted `testOptionsFileRepresentsDefaults` contents to reflect what we wanted to check

Ref: https://github.com/maglnet/ComposerRequireChecker/pull/564#discussion_r1946714547

Ocramius committed:
Render details on all deprecations

Ref: https://github.com/maglnet/ComposerRequireChecker/actions/runs/18938547270/job/54071352115?pr=564

Should help investigating:

```

1) /home/runner/work/ComposerRequireChecker/ComposerRequireChecker/vendor/revolt/event-loop/src/EventLoop/Internal/AbstractDriver.php:182
Increment on non-numeric string is deprecated, use str_increment() instead

Triggered by:

* ComposerRequireCheckerTest\ASTLocator\LocateASTFromFilesTest::testDoNotFailOnParseErrorWithErrorHandler
  /home/runner/work/ComposerRequireChecker/ComposerRequireChecker/test/ComposerRequireCheckerTest/ASTLocator/LocateASTFromFilesTest.php:68

...
```

Ocramius committed:
Ignore deprecation warnings caused by third-party dependencies

Ref: github.com/sebastianbergmann/phpunit/issues/5689

Ocramius committed:
Do not fail on skipped tests

Tests based on the compiled `.phar` are skipped on purpose.

Ocramius committed:
Removed `CheckCommand#initialize()`

It is redundant, and it applies validation before validation is supposed to be used,
according to the parent interface.
